dask/bag/core.py

```diff
@@ -1447,3 +1453,125 @@ def _reduce(binop, sequence, initial=no_default):
         return reduce(binop, sequence, initial)
     else:
         return reduce(binop, sequence)
+
+
+def make_group(k, stage):
+    def h(x):
+        return x[0] // k ** stage % k
+    return h
+
+
+def digit(n, k, base):
+    """
+
+    >>> digit(1234, 0, 10)
+    4
+    >>> digit(1234, 1, 10)
+    3
+    >>> digit(1234, 2, 10)
+    2
+    >>> digit(1234, 3, 10)
+    1
+    """
+    return n // base**k  % base
+
+
+def insert(tup, loc, val):
+    """
+
+    >>> insert(('a', 'b', 'c'), 0, 'x')
+    ('x', 'b', 'c')
+    """
+    L = list(tup)
+    L[loc] = val
+    return tuple(L)
+
+
+def groupby_tasks(b, grouper, hash=hash, max_branch=32,
+                 name='shuffle', token=None):
+    max_branch = max_branch or 32
+    n = b.npartitions
+
+    stages = int(math.ceil(math.log(n) / math.log(max_branch)))
+    if stages > 1:
+        k = int(math.ceil(n ** (1 / stages)))
+    else:
+        k = n
+
+    groups = []
+    splits = []
+    joins = []
+
+    inputs = [tuple(digit(i, j, k) for j in range(stages))
+              for i in range(n)]
+    sinputs = set(inputs)
+
+    b2 = b.map(lambda x: (hash(grouper(x)), x))
+
+    token = token or tokenize(b, hash, max_branch, name)
```

Comment:
Probably don't need `name` in here, but I think you do need `grouper`. Otherwise the output tokens are the same regardless of grouper. Should have a test for that as well.